File: src/R/rsa_encryption.r
library(bit64)
library(gmp)
# bit64::as.integer64!!

get_euler_totient <- function(p, q) {
  bit64::as.integer64((p-1)*(q-1))
}

is_prime <- function(n) {
  if (n<2) {
    return(FALSE)
  }
  
  if (n==2) {
    return(TRUE)
  }
  
  for(i in 2:ceiling(sqrt(n))) {
    if (n %% i == 0) {
      return(FALSE)
    }
  }
  TRUE
}

get_primes <- function(ceil) {
  primes <- c()
  for (i in 2:ceil) {
    if (is_prime(i)) {
      primes <- append(primes, i)
    }
  }
  primes
}

gcd <- function(x,y) {
  r <- x%%y;
  return(ifelse(r, gcd(y, r), y))
}

get_pub_exp <- function(e) {
  primes <- get_primes(e)
  
  for (p in rev(primes)) {
    if (gcd(p, e) == 1) {
      return(p)
    }
  }
}

get_priv_exp <- function(ceil, pub_exp, e) {
  for (num in ceil:1) {
    if ((pub_exp*num) %% e == 1) {
      return(num)
    }
  }
}


modpow <- function(base, exp, modulus) {
  base <- as.bigz(base)
  exp <- as.bigz(pub_key[1])
  modulus <- as.bigz(pub_key[2])
  result <- gmp::powm(base, exp, modulus)
  
  as.numeric(result)
}

#################################################
p <- 223
q <- 229
prod <- p*q

euler_totient <- get_euler_totient(p, q)
public_exp <- get_pub_exp(euler_totient)
private_exp <- get_priv_exp(1e6, public_exp, euler_totient)

public_key <- c(public_exp, prod)
private_key <- c(private_exp, prod)
print(public_key)
print(private_key)

#encr_msg <- encrypt_msg("Hello World!", public_key)

File: src/Ruby/rsa_encryption.rb
def get_euler_totient(p, q)
    return (p-1)*(q-1)
end

def is_prime(n)
    (2..(n/2).floor()).each do |i|
        if n % i == 0
            return false
        end
    end
    return true
end

def get_primes(ceiling)
    primes = []

    (2..ceiling).each do |num|
        if is_prime(num)
            primes.append(num)
        end
    end

    return primes
end

def get_public_exponent(euler_totient)
    primes = get_primes(euler_totient)

    (primes.reverse()).each do |num|
        if num.gcd(euler_totient) == 1
            return num
        end
    end
end

def get_private_exponent(ceiling, public_exp, euler_totient)
    (ceiling).downto(1).each do |num|
        if ((public_exp * num) % euler_totient) == 1
            return num
        end
    end
end

def encrypt_message(message, public_key)
    encr_message = []

    message.each_byte do |ch|
        encr_char = modpow(ch, public_key[0], public_key[1])
        encr_message.append(encr_char)
    end

    return encr_message
end

def decrypt_message(encr_message, private_key)
    decr_chars = []

    encr_message.each do |encr_ch|
        ascii = modpow(encr_ch, private_key[0], private_key[1])
        decr_chars.append(ascii.chr)
    end

    return decr_chars.join()
end

def modpow(base, exp, mod)
    res = 1

    (1..exp).each do |i|
        res = (res * base) % mod
    end

    return res
end

def main()
    num_p = 223
    num_q = 229
    num_product = num_p * num_q

    euler_totient = get_euler_totient(num_p, num_q)
    public_exp = get_public_exponent(euler_totient)
    private_exp = get_private_exponent(1_000_000, public_exp, euler_totient)

    public_key = [public_exp, num_product]
    private_key = [private_exp, num_product]
    # print(public_key, private_key)

    encr_message = encrypt_message("Hello World!", public_key)
    decr_message = decrypt_message(encr_message, private_key)
    print(decr_message)
end

main()
